rPackages.littler: fix build

ckauhaus opened the pull request:
###### Motivation for this change

The littler build was completely broken. Fix littler by providing necessary deps and moving build artefacts into bin, man, doc.

Usage:

```
nix-env -iA nixpkgs.rPackages.littler
# or put rPackages.littler into environment.systemPacakges
man r
r
```

Provides an alias `lr` for non-case-preserving filesystems (think Darwin) as proposed by the package author.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after) -> N/A (build was broken before)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Commit:
rPackages.littler: fix build

littler gives a wrapped called `r` (or `lr` for non-case-preserving
systems like Darwin) which works as shebang or pipe target.

The build was completely broken before (missing deps).

Symlink executables and manpage into standard locations so that this
packages also works in `environment.systemPackages`.

peti commented on pkgs/development/r-modules/default.nix:
```diff
@@ -950,6 +950,18 @@ let
+    littler = old.littler.overrideAttrs (attrs: with pkgs; {
+      buildInputs = [ pcre lzma zlib bzip2 icu which ] ++ attrs.buildInputs;
+      propagatedBuildInputs = [ which ] ++ attrs.propagatedBuildInputs;
+      outputs = [ "out" "man" "doc" ];
```
No other R package provides those outputs, so don't think it's a good idea to add them here because nobody will expect those outputs to exist.

ckauhaus replied:
Do you think I should just leave split-outputs out (put everything into $out) or that this package should rather do to top-level `all-packages.nix`? It differs from most other R packages in the way that it provides a separate entry point into the R ecosystem.

peti commented on pkgs/development/r-modules/default.nix:
```diff
@@ -950,6 +950,18 @@ let
 
+    littler = old.littler.overrideAttrs (attrs: with pkgs; {
+      buildInputs = [ pcre lzma zlib bzip2 icu which ] ++ attrs.buildInputs;
+      propagatedBuildInputs = [ which ] ++ attrs.propagatedBuildInputs;
```
Why does this package have to propagate `which`?

ckauhaus replied:
Because it won't build otherwise. Example error output in case this is left out:
```
** testing if installed package can be loaded
Error: package or namespace load failed for 'littler':
 .onAttach failed in attachNamespace() for 'littler', details:
  call: system(paste(which, shQuote(names[i])), intern = TRUE, ignore.stderr = TRUE)
  error: error in running command
Error: loading failed
```

Taking it one level back it seems that access to `which` is actually a core dependency of R. There are 15+ package overrides present in https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/r-modules/default.nix that put `which` into propagatedBuildInputs on a per-package basis. Perhaps it would even cleaner to move `which` from `buildInputs` to `propagatedBuildInputs` in https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/science/math/R/default.nix#L18. Thoughts?

peti commented on pkgs/development/r-modules/default.nix:
```diff
@@ -950,6 +950,18 @@ let
+      propagatedBuildInputs = [ which ] ++ attrs.propagatedBuildInputs;
+      outputs = [ "out" "man" "doc" ];
+      postInstall = ''
+        install -D inst/bin/r $out/bin/r
```
The install process copies dozens of files into `$out`, albeit into strange locations. Now, this code will add even more files, some of them clearly duplicates. I think it would be better to clean up the generated output, i.e. remove files that aren't needed and move others into the right place.

ckauhaus replied:
Yeah, I'll clean it up. But we should probably decide upon the questions at top.

ckauhaus commented:
@peti Second attempt.

Things changed:
- `which` is now a build-time not run-time dependency
- removed multiple outputs
- de-duplicated $out: bin and man now symlink into their R library locations
- clean up: removed script-tests since they aren't useful here
- left docs, examples and html in place since most other R packages do the same.

Now acceptable?
